[GStreamer][MediaStream] Video resizing fixes

philn opened the pull request:
#### 8a44e02b866aad92a8c613ead7ed7d40daf45a1f
<pre>
[GStreamer][MediaStream] Video resizing fixes
<a href="https://bugs.webkit.org/show_bug.cgi?id=243638">https://bugs.webkit.org/show_bug.cgi?id=243638</a>

Reviewed by Xabier Rodriguez-Calvar.

Resizing a media track from JS triggers a call to RealtimeMediaSource::setSize() but that doesn&apos;t
affect the mock video source per-se because it is wrapped in a RealtimeVideoSource. On Apple
platforms the RealtimeVideoSource actually resizes the video frame, but keeps the captureSize
un-modified. So the RealtimeVideoSource::adaptVideoFrame() method is also now implemented for
GStreamer ports.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::setBufferFields):
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
(WebCore::VideoFrameGStreamer::resizeTo):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::adaptVideoFrame):
(WebCore::RealtimeVideoSource::videoFrameAvailable):
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h:

Canonical link: <a href="https://commits.webkit.org/253289@main">https://commits.webkit.org/253289@main</a>
</pre>

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
```diff
@@ -30,6 +30,16 @@
+    GST_BUFFER_DTS(buffer) = GST_BUFFER_PTS(buffer) = toGstClockTime(presentationTime);
+
+    auto duration = 1_s / frameRate;
+    GST_BUFFER_DURATION(buffer) = toGstClockTime(MediaTime::createWithDouble(duration.seconds()));
```
I've seen you using this pattern a couple of times already. You might want to overload `toGstClockTime` to take `const Seconds &`

philn replied:
Good idea!

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
```diff
@@ -134,6 +135,46 @@ VideoFrameGStreamer::VideoFrameGStreamer(const GRefPtr<GstSample>& sample, const
+        GstMappedFrame outputFrame(outputBuffer.get(), outputInfo, GST_MAP_WRITE);
+        gst_video_converter_frame(converter.get(), inputFrame.get(), outputFrame.get());
+    }
+    gst_buffer_add_video_meta(outputBuffer.get(), GST_VIDEO_FRAME_FLAG_NONE, format, width, height);
```
I don't think this meta is required, since it's useful when the strides or alignment aren't the common ones for the format.

philn committed:
[GStreamer][MediaStream] Video resizing fixes
https://bugs.webkit.org/show_bug.cgi?id=243638

Reviewed by Xabier Rodriguez-Calvar.

Resizing a media track from JS triggers a call to RealtimeMediaSource::setSize() but that doesn't
affect the mock video source per-se because it is wrapped in a RealtimeVideoSource. On Apple
platforms the RealtimeVideoSource actually resizes the video frame, but keeps the captureSize
un-modified. So the RealtimeVideoSource::adaptVideoFrame() method is also now implemented for
GStreamer ports.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::setBufferFields):
(WebCore::VideoFrameGStreamer::createFromPixelBuffer):
(WebCore::VideoFrameGStreamer::resizeTo):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
* Source/WebCore/platform/mediastream/RealtimeVideoSource.cpp:
(WebCore::RealtimeVideoSource::adaptVideoFrame):
(WebCore::RealtimeVideoSource::videoFrameAvailable):
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h:

Canonical link: https://commits.webkit.org/253289@main
